src/LowerWarpShuffles.cpp

```diff
@@ -0,0 +1,760 @@
+        // If it was written outside this if clause but read inside of
+        // it, we need to hoist it.
+        if (starts_with(op->name, "llvm.nvvm.shfl.") &&
+            !expr_uses_vars(op, stored_to)) {
+            string name = unique_name('t');
+            lifted_lets.push_back({name, op});
+            return Variable::make(op->type, name);
+        } else {
+            return IRMutator2::visit(op);
+        }
+    }
+
+    template<typename ExprOrStmt, typename LetOrLetStmt>
+    ExprOrStmt visit_let(const LetOrLetStmt *op) {
+        Expr value = mutate(op->value);
+        ExprOrStmt body = mutate(op->body);
+
+        // If any of the lifted expressions use this, we also need to
+        // lift this.
+        bool should_lift = false;
+        for (const auto &p : lifted_lets) {
+            should_lift |= expr_uses_var(p.second, op->name);
+        }
+
+        if (should_lift) {
+            lifted_lets.push_back({op->name, value});
+            return body;
+        } else {
+            return LetOrLetStmt::make(op->name, value, body);
+        }
+    }
+
+    Expr visit(const Let *op) override {
+        return visit_let<Expr>(op);
+    }
+
+    Stmt visit(const LetStmt *op) override {
+        return visit_let<Stmt>(op);
+    }
+
+    Stmt visit(const Store *op) override {
+        stored_to.push(op->name, 0);
+        return IRMutator2::visit(op);
+    }
+public:
+    Stmt rewrap(Stmt s) {
+        while (!lifted_lets.empty()) {
+            const pair<string, Expr> &p = lifted_lets.back();
+            s = LetStmt::make(p.first, p.second, s);
+            lifted_lets.pop_back();
+        }
+        return s;
+    }
+};
+
+// The destination *and source* for warp shuffles must be active
+// threads, or the value is undefined, so we want to lift them out of
+// if statements.
+class HoistWarpShuffles : public IRMutator2 {
+    using IRMutator2::visit;
+
+    Stmt visit(const IfThenElse *op) override {
+        // Move all Exprs that contain a shuffle out of the body of
+        // the if.
+        Stmt then_case = mutate(op->then_case);
+        Stmt else_case = mutate(op->else_case);
+
+        HoistWarpShufflesFromSingleIfStmt hoister;
+        then_case = hoister.mutate(then_case);
+        else_case = hoister.mutate(else_case);
+        Stmt s = IfThenElse::make(op->condition, then_case, else_case);
+        return hoister.rewrap(s);
+    }
+};
+
+class HasLaneLoop : public IRVisitor {
+    using IRVisitor::visit;
+
+    void visit(const For *op) {
+        result = result || op->for_type == ForType::GPULane;
+        IRVisitor::visit(op);
+    }
+public:
+    bool result = false;
+};
+
+bool has_lane_loop(Stmt s) {
+    HasLaneLoop l;
+    s.accept(&l);
+    return l.result;
+}
+
+class LowerWarpShufflesInEachKernel : public IRMutator2 {
+    using IRMutator2::visit;
+
+    Stmt visit(const For *op) override {
+        if (op->device_api == DeviceAPI::CUDA && has_lane_loop(op)) {
```

Comment:
Doesn't LowerWarpShuffles also do things if there's an allocation between thread loops of constant extent?

Reply:
That's actually illegal. You can only have allocations between the blocks and threads loop, and between the threads and lanes loop